### ISSUE TITLE
cl_event misuse crashes the runtime

### ISSUE THREAD

**Issue by MathiasMagnus** [pocl/pocl#703]
There have been a few misuses of cl_event in an application of mine which "unfortunately" the Windows runtime of AMD handled just fine. However, when moving the code to test on Linux, it crashed on multiple erronous usage of cl_event entities instead of signaling errors.

- Calling `clReleaseEvent()` on an event that already has been released will trigger an assertion error [here](https://github.com/pocl/pocl/blob/release_1_1/lib/CL/clReleaseEvent.c#L35) which is then relayed to a generic [release object](https://github.com/pocl/pocl/blob/release_1_1/lib/CL/pocl_cl.h#L112) macro and lock mechanism that [fires](https://github.com/pocl/pocl/blob/release_1_1/lib/CL/pocl_cl.h#L101) the actual assertion in case ref count has already hit zero. Correct behavior would be to return `CL_INVALID_EVENT` error as mandated by the [spec](https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clReleaseEvent.html).
- Waiting on such a released event fires `The futex facility returned an unexpected error code` from somewhere within the runtime, instead of the same `CL_INVALID_EVENT` as yet again mandated by the [spec](https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clWaitForEvents.html).
- Trying to obtain profiling info via `clGetEventProfilingInfo()` results in the program aborting, which is also not the nicest way to handle things.

**Comment by pjaaskel** [comment]
Right, this has been discussed before in some another issue: Returning *CL_INVALID_object* due to too many releases requires keeping a white list of known valid objects in the context, or keeping the objects alive (do not actually free them) just to denote their invalidity (with an internal flag indicating so), both of which are problematic, therefore it's not currently implemented. The list approach still does not fix the case if the same object (with same address) is reallocated to a new event (then the buggy client code calls the functions on wrong data) and the another just leaks memory. Any attempts to fix this cleanly are appreciated.
